File: JsonSchemaAndStructs/KiteConnect-Kotlin-Structs/mf_instruments.kt
// To parse the JSON, install kotlin's serialization plugin and do:
//
// val json          = Json(JsonConfiguration.Stable)
// val mFInstruments = json.parse(MFInstruments.serializer(), jsonString)

package MfInstruments

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*

typealias MFInstruments = JsonArray<MFInstrument>

@Serializable
data class MFInstrument (
    val amc: Amc? = null,

    @SerialName("dividend_type")
    val dividendType: DividendType? = null,

    @SerialName("last_price")
    val lastPrice: Double? = null,

    @SerialName("last_price_date")
    val lastPriceDate: String? = null,

    @SerialName("minimum_additional_purchase_amount")
    val minimumAdditionalPurchaseAmount: Long? = null,

    @SerialName("minimum_purchase_amount")
    val minimumPurchaseAmount: Long? = null,

    @SerialName("minimum_redemption_quantity")
    val minimumRedemptionQuantity: Double? = null,

    val name: String? = null,
    val plan: Plan? = null,

    @SerialName("purchase_allowed")
    val purchaseAllowed: Long? = null,

    @SerialName("purchase_amount_multiplier")
    val purchaseAmountMultiplier: Long? = null,

    @SerialName("redemption_allowed")
    val redemptionAllowed: Long? = null,

    @SerialName("redemption_quantity_multiplier")
    val redemptionQuantityMultiplier: Double? = null,

    @SerialName("scheme_type")
    val schemeType: SchemeType? = null,

    @SerialName("settlement_type")
    val settlementType: SettlementType? = null,

    val tradingsymbol: String? = null
)

@Serializable
enum class Amc(val value: String) {
    BirlaSunLifeMutualFundMF("BirlaSunLifeMutualFund_MF");

    companion object : KSerializer<Amc> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("MfInstruments.Amc", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): Amc = when (val value = decoder.decodeString()) {
            "BirlaSunLifeMutualFund_MF" -> BirlaSunLifeMutualFundMF
            else                        -> throw IllegalArgumentException("Amc could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: Amc) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class DividendType(val value: String) {
    Growth("growth"),
    Payout("payout");

    companion object : KSerializer<DividendType> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("MfInstruments.DividendType", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): DividendType = when (val value = decoder.decodeString()) {
            "growth" -> Growth
            "payout" -> Payout
            else     -> throw IllegalArgumentException("DividendType could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: DividendType) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class Plan(val value: String) {
    Direct("direct"),
    Regular("regular");

    companion object : KSerializer<Plan> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("MfInstruments.Plan", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): Plan = when (val value = decoder.decodeString()) {
            "direct"  -> Direct
            "regular" -> Regular
            else      -> throw IllegalArgumentException("Plan could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: Plan) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class SchemeType(val value: String) {
    Balanced("balanced"),
    Debt("debt"),
    Equity("equity"),
    Fof("fof"),
    Liquid("liquid");

    companion object : KSerializer<SchemeType> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("MfInstruments.SchemeType", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): SchemeType = when (val value = decoder.decodeString()) {
            "balanced" -> Balanced
            "debt"     -> Debt
            "equity"   -> Equity
            "fof"      -> Fof
            "liquid"   -> Liquid
            else       -> throw IllegalArgumentException("SchemeType could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: SchemeType) {
            return encoder.encodeString(value.value)
        }
    }
}

@Serializable
enum class SettlementType(val value: String) {
    T1("T1"),
    T3("T3"),
    T4("T4"),
    T6("T6");

    companion object : KSerializer<SettlementType> {
        override val descriptor: SerialDescriptor get() {
            return PrimitiveSerialDescriptor("MfInstruments.SettlementType", PrimitiveKind.STRING)
        }
        override fun deserialize(decoder: Decoder): SettlementType = when (val value = decoder.decodeString()) {
            "T1" -> T1
            "T3" -> T3
            "T4" -> T4
            "T6" -> T6
            else -> throw IllegalArgumentException("SettlementType could not parse: $value")
        }
        override fun serialize(encoder: Encoder, value: SettlementType) {
            return encoder.encodeString(value.value)
        }
    }
}

File: JsonSchemaAndStructs/KiteConnect-Rust-Structs/mf_instruments.rs
// Example code that deserializes and serializes the model.
// extern crate serde;
// #[macro_use]
// extern crate serde_derive;
// extern crate serde_json;
//
// use generated_module::[object Object];
//
// fn main() {
//     let json = r#"{"answer": 42}"#;
//     let model: [object Object] = serde_json::from_str(&json).unwrap();
// }

extern crate serde_derive;

pub type MfInstruments = Vec<MfInstrument>;

#[derive(Debug, Serialize, Deserialize)]
pub struct MfInstrument {
    pub amc: Option<Amc>,
    pub dividend_type: Option<DividendType>,
    pub last_price: Option<f64>,
    pub last_price_date: Option<String>,
    pub minimum_additional_purchase_amount: Option<i64>,
    pub minimum_purchase_amount: Option<i64>,
    pub minimum_redemption_quantity: Option<f64>,
    pub name: Option<String>,
    pub plan: Option<Plan>,
    pub purchase_allowed: Option<i64>,
    pub purchase_amount_multiplier: Option<i64>,
    pub redemption_allowed: Option<i64>,
    pub redemption_quantity_multiplier: Option<f64>,
    pub scheme_type: Option<SchemeType>,
    pub settlement_type: Option<SettlementType>,
    pub tradingsymbol: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Amc {
    #[serde(rename = "BirlaSunLifeMutualFund_MF")]
    BirlaSunLifeMutualFundMf,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum DividendType {
    #[serde(rename = "growth")]
    Growth,
    #[serde(rename = "payout")]
    Payout,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Plan {
    #[serde(rename = "direct")]
    Direct,
    #[serde(rename = "regular")]
    Regular,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum SchemeType {
    #[serde(rename = "balanced")]
    Balanced,
    #[serde(rename = "debt")]
    Debt,
    #[serde(rename = "equity")]
    Equity,
    #[serde(rename = "fof")]
    Fof,
    #[serde(rename = "liquid")]
    Liquid,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum SettlementType {
    T1,
    T3,
    T4,
    T6,
}
